### PR TITLE
NewSnapshotReader Nil Error

### PR DESCRIPTION
**Pull request by jcastrence**
#### Type of change

- Improvement (improvement to code, performance, etc)

#### Description

- Edited NewSnapshotReader so it now logs if the passed in data file does not exist

#### Related issues

- [FAB-18469](https://jira.hyperledger.org/browse/FAB-18469)

### CODE REVIEW

**Review by C0rWin — DISMISSED** [review]

File: core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot.go [hunk 1]
```diff
@@ -441,9 +441,12 @@ func newSnapshotReader(dir, dataFileName, metadataFileName string) (*snapshotRea
 	dataFilePath := filepath.Join(dir, dataFileName)
 	metadataFilePath := filepath.Join(dir, metadataFileName)
 	exist, _, err := fileutil.FileExists(dataFilePath)
-	if err != nil || !exist {
+	if err != nil {
 		return nil, err
 	}
+	if !exist {
+		return nil, errors.New("file does not exist")
```
Comment (manish-sethi) [inline core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot.go]:
```suggestion
		return nil, errors.Errorf("file [%s] does not exist", dataFilePath)
```

File: core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot_test.go [hunk 2]
```diff
@@ -363,6 +363,9 @@ func TestSnapshotReaderNextFunction(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, "", retrievedNs)
 	require.Nil(t, retrievedSr)
+
+	_, err = newSnapshotReader("test-dir", "test-data-filename", "test-metadata-filename")
+	require.EqualError(t, err, "file does not exist")
```
Comment (manish-sethi) [inline core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot_test.go]: This test is misplaced, under `TestSnapshotReaderNextFunction`. Moving this under `TestSnapshotImportErrorPropagation` makes a better fit.

**Comment by manish-sethi** [comment]
My suggestion was to simply add the following code in the test file at [line 535](https://github.com/hyperledger/fabric/blob/main/core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot_test.go#L535).

```
		t.Run("error_missing_data_file_"+f, func(t *testing.T) {
			init()
			defer cleanup()

			dataFile := filepath.Join(snapshotDir, f)
			require.NoError(t, os.Remove(dataFile))
			err := dbEnv.GetProvider().ImportFromSnapshot(
				generateLedgerID(t), version.NewHeight(10, 10), snapshotDir)
			require.Contains(t, err.Error(), fmt.Sprintf("file [%s] does not exist", dataFile))
		})
```

**Comment by jcastrence** [comment]
> My suggestion was to simply add the following code in the test file at [line 535](https://github.com/hyperledger/fabric/blob/main/core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot_test.go#L535).
> 
> ```
> 		t.Run("error_missing_data_file_"+f, func(t *testing.T) {
> 			init()
> 			defer cleanup()
> 
> 			dataFile := filepath.Join(snapshotDir, f)
> 			require.NoError(t, os.Remove(dataFile))
> 			err := dbEnv.GetProvider().ImportFromSnapshot(
> 				generateLedgerID(t), version.NewHeight(10, 10), snapshotDir)
> 			require.Contains(t, err.Error(), fmt.Sprintf("file [%s] does not exist", dataFile))
> 		})
> ```

Right, however the changes I've made to snapshot.go have caused some of the existing test cases to fail since they are missing either public state or private state hashes. Because of the addition of a new error being thrown if files are missing, they will fail since they are not expecting this new error. I thought that it'd be best to then test these test cases with missing files separately which ends up covering the addition you suggested. What are your thoughts on this?

**Comment by manish-sethi** [comment]
> What are your thoughts on this?

My suggestion would be to fix the input to the existing test.

**Comment by denyeart** [comment]
Upon further discussion the nil response is expected in some scenarios (such as no private state data) so we can simply leave the functionality as-is and improve the comments and logging to make the intended use more clear.

File: core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot.go [hunk 3]
```diff
@@ -437,12 +438,17 @@ type snapshotReader struct {
...
 	dataFilePath := filepath.Join(dir, dataFileName)
 	metadataFilePath := filepath.Join(dir, metadataFileName)
 	exist, _, err := fileutil.FileExists(dataFilePath)
-	if err != nil || !exist {
-		return nil, err
+	if err != nil {
+		return nil, errors.WithMessage(err, "error while checking if data file exists")
+	}
+	if !exist {
+		log.Printf("Snapshot data file name does not exist")
```
Comment (manish-sethi) [inline core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot.go]:
```suggestion
	        logger.Infow("Data file does not exist. Nothing to be done.", "filepath", dataFilePath)
```

**Review by manish-sethi — APPROVED** [review]

**Pull request merged**